### ISSUE TITLE
Reduce scope to what is strictly needed in `levelup`

### ISSUE THREAD

**Issue by vweevers** [Level/deferred-leveldown#90]
_TLDR: thumbs up if you agree that deferring operations is only useful while the db is opening._

I'm working on [`getMany()`](https://github.com/Level/community/issues/101). Which is a new [`abstract-leveldown`](https://github.com/Level/abstract-leveldown/pull/381) and [`levelup`](https://github.com/Level/levelup/pull/725) method that gives `abstract-leveldown` the responsibility of checking whether the db is open when `getMany()` is called. As an experiment for https://github.com/Level/community/issues/58. But I'm running into an inconsistency that's particularly problematic in `subleveldown` (which has to deal with two instances of `deferred-leveldown` and `levelup`, and is a [test bed](https://github.com/Level/subleveldown/blob/1f9d550672d0550a9af001353edd24d0f5ee5153/test/index.js#L70-L89) for API parity).

`deferred-leveldown` and `levelup` are inconsistent in their [deferredOpen](https://github.com/Level/supports#deferredopen-boolean) behavior: the former allows operations while the db has any status, the latter only if 'opening' or 'open'. For example, if the status is 'closed', `deferred-leveldown` will put operations in its queue. But in practice we don't get there because `levelup` will yield a [`new ReadError('Database is not open')`](https://github.com/Level/levelup/blob/795b393dc7e2b9ed418c55483082d4c04c4f06f6/lib/levelup.js#L305-L308).

@ralphtheninja @juliangruber Given that the order of [initialization](https://github.com/Level/levelup/blob/795b393dc7e2b9ed418c55483082d4c04c4f06f6/lib/levelup.js#L61-L65)  in `levelup` is:

- Wrap input `db` with `deferred-leveldown`
- Start opening it (changing its status from 'new' to 'opening')
- Start accepting operations (after returning from constructor)

I propose the following changes:

- `deferred-leveldown` will only defer operations if the status of its underlying db is 'opening'
- Otherwise it will yield `new Error('Database is not open')`.

This is a breaking change for `deferred-leveldown`, but semver-patch for `levelup`.

See also https://github.com/Level/levelup/issues/710#issuecomment-753611245.

**Comment by juliangruber** [comment]
> * `deferred-leveldown` will only defer operations if the status of its underlying db is 'opening'

I think there are some more cases we could discuss. Say the db was closed, should another operation cause it to reopen? Or it could even be currently closing and a consumer could expect the next operation to reopen it.

If after closing a db the user wants to use the deferred-leveldown behavior again, they need to ensure they called db.open(), so that it's in the opening state, right? I think that makes sense to me. Just want to check it with you

**Comment by vweevers** [comment]
> If after closing a db the user wants to use the deferred-leveldown behavior again, they need to ensure they called db.open(), so that it's in the opening state, right? I think that makes sense to me. Just want to check it with you

Right. The way I see it, deferredOpen is a convenience to avoid having to open the db. It goes away if the user explicitly closes the db. It then becomes the user's responsibility to explicitly open the db again. After which the convenience is back.

I reckon in most if not all cases, if the user is closing the db, they're not expecting to do further work on the db, and scheduling operations _should_ be an error.

**Comment by vweevers** [comment]
Which is to say, yielding `new Error('Database is not open')` is also a convenience :)

**Comment by ralphtheninja** [comment]
`deferred-leveldown` feels a bit hacky to me. I just want to entertain the idea of removing it in the long run. My gut feeling tells me we're having to fight it all the time and it's difficult to see what's going on. Might be a lot of work.

**Comment by vweevers** [comment]
Oh, I _definitely_ want to remove it. Just gradually.

**Comment by juliangruber** [comment]
I want to remove it as well. I think we mostly had it because of callback inconvenience, now that we can `await db.open()` there's no big advantage any more

**Comment by juliangruber** [comment]
> I reckon in most if not all cases, if the user is closing the db, they're not expecting to do further work on the db, and scheduling operations _should_ be an error.

I think you're right. The easiest path for them is to dispose of the db completely after closing it, and then starting fresh. So only treating the `opening` state makes sense 👍

**Comment by vweevers** [comment]
To clarify, I want to remove `deferred-leveldown`, but not necessarily deferredOpen (as a feature). It does have a major benefit on sublevels for example. Don't want to have to do:

```js
await db.open()
const sub = require('subleveldown')(db)
await sub.open()
```

That, plus widespread usage of deferredOpen, and that we may not be seeing other benefits, is why I want to remove it gradually. We may find a new place for it, or that opening sublevels isn't necessary, or... We'll see.

**Comment by juliangruber** [comment]
Why don't you want to have to do that? That code looks fine to me. Async operations are cheap now, I think we can remove deferredOpen altogether

**Comment by vweevers** [comment]
Asynchronicity is viral. Let's say a module takes a db as input, for example `function myIndexer(db)` which internally creates sublevels to store indexed data - or just to have a fresh db instance that can safely be monkeypatched. And it can't write to those sublevels until a `await sub.open()`. Assume that this module wants to intercept writes on the input db, in order to update indexes. Then, because of the required async initialization, it must change its signature to `async function myIndexer(db)`.

**Comment by juliangruber** [comment]
Pseudo code:

```js
const myIndexer = db => {
  db.put = async (key, value) => {
    await sub.open()
    // ...
  }
}
```

Wouldn't something like this work? By moving the async operation into the db hooks

**Comment by vweevers** [comment]
That just moves the problem. I think in userland you shouldn't have to worry about open/close state; saves boilerplate and prevents race issues.

deferredOpen isn't necessarily a complex feature (that should therefore be removed). The problem I have with the current implementation is that the feature is incomplete without `levelup`. Because it's `levelup` that auto-opens the db (and emits the events that are needed for logic like `db.once('open', cb)` if we wanted to implement it elsewhere). So:

```js
require('levelup')(db).supports.deferredOpen // true
require('deferred-leveldown')(db).supports.deferredOpen // false!
```
